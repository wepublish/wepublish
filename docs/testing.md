# Automated Testing

Automated testing is a crucial aspect of software development that helps to ensure the quality of a codebase by
automating the testing process.

## API Unit Testing

API unit testing is a type of automated testing that focuses on testing individual units of code that make up an API.

### Setting up the Testing Environment

To set up the testing environment, you need to ensure that you have the following entry in your `jest.config.ts` file: `globalSetup: '<rootDir>/setup-database.js'`. You can find an example of this in the `libs/membership/api/src` folder.

In the `setup-database.js` file, you should have code that looks like the following:

```typescript
import execa from 'execa'

export default async () => {
    const databaseUrl = 'postgresql://postgres@localhost:5432/wepublish_test?schema=public'
    process.env.DATABASE_URL = databaseUrl
    process.env.TZ = 'UTC'

    await execa(`npx`, ['prisma', 'migrate', 'reset', '--force'])
}
```
This code sets up a test database called `wepublish_test` and resets it each time you run a test.

### Running Tests
To run tests, you need to ensure that all your changes to the `schema.prisma` file are reflected in a migration. 
You can do this by running `prisma migrate dev --name name_of_my_migration`. Otherwise, your test database may 
rely on an incomplete schema.
Also, make sure, you have a database up and running. You may want run `docker-compose up database`

Once you have done this, you should be able to run a single test using the command `npx jest -t 'my test'`

### Writing Tests with Fabbrica
When it comes to writing tests, we highly recommend using [Fabbrica](https://github.com/Quramy/prisma-fabbrica).
With Fabbrica, you can easily generate database entries using the `.create()` method or model instances using
the `.build()` method, saving you time and effort.

For examples of how to use Fabbrica in your tests, check out the `subscription-flow.controller.spec.ts` file 
located at `libs/membership/api/src/lib/subscription-flow`. It's a great resource to help you get started with writing
tests that utilize Fabbrica.

### ! Do not forget adding tests to the pipeline !
In order to run your tests within the pipeline, you have to add an entry in the `package.json` in the root folder.
To test, if it works, run `nx test` from your cli. Your specs should now be tested.


## Snapshots
It might happen that your tests are failing. Try to update your snapshots
by running (eventually in the package in question) `npm run test -- -u`

For more information: [https://jestjs.io/docs/cli#--updatesnapshot](https://jestjs.io/docs/cli#--updatesnapshot)