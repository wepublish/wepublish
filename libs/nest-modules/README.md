# nest-modules

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build nest-modules` to build the library.

## Running unit tests

Run `nx test nest-modules` to execute the unit tests via [Jest](https://jestjs.io).
