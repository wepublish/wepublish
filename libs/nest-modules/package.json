{
  "name": "@wepublish/nest-modules",
  "version": "9.2.0",
  "description": "Nestjs API modules for we.publish.",
  "keywords": [
    "nestjs",
    "publishing",
    "media",
    "newspaper"
  ],
  "homepage": "https://github.com/wepublish/wepublish",
  "bugs": {
    "url": "https://github.com/wepublish/wepublish/issues",
    "email": "development@wepublish.ch"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/wepublish/wepublish.git",
    "directory": "libs/nest-modules"
  },
  "license": "MIT",
  "publishConfig": {
    "directory": "../../dist/libs/nest-modules"
  }
}
