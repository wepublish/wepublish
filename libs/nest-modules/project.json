{
  "name": "nest-modules",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/nest-modules/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/libs/nest-modules",
        "tsConfig": "libs/nest-modules/tsconfig.lib.json",
        "packageJson": "libs/nest-modules/package.json",
        "main": "libs/nest-modules/src/index.ts",
        "assets": ["libs/nest-modules/*.md"]
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/nest-modules/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/nest-modules/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "tags": []
}
