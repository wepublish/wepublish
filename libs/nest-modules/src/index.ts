export * from './lib/api.module'
export * from './lib/database/prisma.module'
export * from './lib/add-metadata.decorator'
