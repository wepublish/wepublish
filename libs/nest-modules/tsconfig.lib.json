{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../../dist/out-tsc",
    "declaration": true,
    "types": ["node"],
    "target": "es6"
  },
  "include": ["src/**/*.ts", "../authentication/api/src/lib/one-of.decorator.ts"],
  "exclude": ["jest.config.ts", "src/**/*.spec.ts", "src/**/*.test.ts"]
}
