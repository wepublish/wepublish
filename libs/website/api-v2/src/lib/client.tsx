import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {ComponentType, memo, useMemo} from 'react'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

export const getV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({
    uri: `${apiUrl}/v2`,
    batchMax: 5,
    batchInterval: 20
  })
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV2ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV2ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV2ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
