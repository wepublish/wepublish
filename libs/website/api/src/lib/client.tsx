import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

import {ComponentType, memo, useMemo} from 'react'

export const getV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({uri: `${apiUrl}/v1`, batchMax: 5, batchInterval: 20})
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV1ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV1ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV1ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
