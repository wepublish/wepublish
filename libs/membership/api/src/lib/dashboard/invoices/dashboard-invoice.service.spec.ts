import {Test, TestingModule} from '@nestjs/testing'
import {DashboardInvoiceService} from './dashboard-invoice.service'
import {PrismaModule} from '@wepublish/nest-modules'
import {PrismaClient} from '@prisma/client'

describe('DashboardInvoiceService', () => {
  let service: DashboardInvoiceService
  let prisma: PrismaClient

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [PrismaModule],
      providers: [DashboardInvoiceService]
    }).compile()

    prisma = module.get<PrismaClient>(PrismaClient)
    service = module.get<DashboardInvoiceService>(DashboardInvoiceService)
  })

  test('should be defined', () => {
    expect(service).toBeDefined()
  })

  test('should get the revenue of a month', async () => {
    const mockValue = Promise.resolve([
      {
        id: 'cld33096g0486tnpxz8espm6e',
        createdAt: '2023-01-19T12:41:05.656Z',
        modifiedAt: '2023-01-19T12:48:13.261Z',
        dueAt: '2023-01-19T12:41:05.640Z',
        paidAt: '2023-01-19T12:48:13.188Z',
        canceledAt: null,
        subscriptionID: 'cld33094l0464tnpx010w5lt6',
        items: [
          {
            quantity: 1,
            amount: 50
          },
          {
            quantity: 2,
            amount: 50
          }
        ],
        subscription: {memberPlan: {name: 'Foo Subscription'}}
      },
      {
        id: 'cld32r1kb0225qppx8ptpwukz',
        createdAt: '2023-01-19T12:33:55.883Z',
        modifiedAt: '2023-01-19T12:34:02.015Z',
        dueAt: '2023-01-19T12:33:55.859Z',
        paidAt: '2023-01-19T12:34:01.944Z',
        canceledAt: null,
        manuallySetAsPaidByUserId: 'cld2zv5hl0096hkpx2hayynif',
        items: [
          {
            quantity: 1,
            amount: 100
          },
          {
            quantity: 2,
            amount: 50
          }
        ],
        subscription: {memberPlan: {name: 'Foo Subscription'}}
      }
    ])
    const start = new Date('2023-01-01')
    const end = new Date('2023-02-01')
    const mockFunction = jest.spyOn(prisma.invoice, 'findMany').mockReturnValue(mockValue as any)

    const result = await service.revenue(start, end)
    expect(result).toMatchSnapshot()
    expect(mockFunction.mock.calls[0][0]).toMatchSnapshot()
  })

  test('should get the expected revenue of a month', async () => {
    const mockValue = Promise.resolve([
      {
        dueAt: '2023-01-19T12:27:23.265Z',
        paidAt: null,
        canceledAt: '2023-01-19T12:27:46.690Z',
        items: [
          {
            quantity: 1,
            amount: 50
          },
          {
            quantity: 2,
            amount: 50
          }
        ],
        subscription: {memberPlan: {name: 'Foo Subscription'}}
      },
      {
        dueAt: '2023-01-19T12:33:55.859Z',
        paidAt: '2023-01-19T12:34:01.944Z',
        canceledAt: null,
        items: [
          {
            quantity: 1,
            amount: 100
          },
          {
            quantity: 2,
            amount: 50
          }
        ],
        subscription: {memberPlan: {name: 'Foo Subscription'}}
      },
      {
        id: 'cld33096g0486tnpxz8espm6e',
        createdAt: '2023-01-19T12:41:05.656Z',
        modifiedAt: '2023-01-19T12:48:13.261Z',
        mail: 'foo@wepublish.ch',
        dueAt: '2023-01-19T12:41:05.640Z',
        description: 'Membership from 2023-01-19T12:41:05.640Z for Merkli',
        paidAt: '2023-01-19T12:48:13.188Z',
        canceledAt: null,
        manuallySetAsPaidByUserId: 'cld2zv5hl0096hkpx2hayynif',
        items: [
          {
            quantity: 1,
            amount: 500
          },
          {
            quantity: 2,
            amount: 250
          }
        ],
        subscription: {memberPlan: {name: 'Foo Subscription'}}
      }
    ])
    const start = new Date('2023-01-01')
    const end = new Date('2023-02-01')
    const mockFunction = jest.spyOn(prisma.invoice, 'findMany').mockReturnValue(mockValue as any)

    const result = await service.expectedRevenue(start, end)
    expect(result).toMatchSnapshot()
    expect(mockFunction.mock.calls[0][0]).toMatchSnapshot()
  })
})
