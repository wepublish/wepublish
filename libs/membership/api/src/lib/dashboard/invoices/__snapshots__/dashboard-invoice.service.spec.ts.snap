// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardInvoiceService should get the expected revenue of a month 1`] = `
[
  {
    "amount": 150,
    "dueAt": "2023-01-19T12:27:23.265Z",
    "memberPlan": "Foo Subscription",
    "paidAt": undefined,
  },
  {
    "amount": 200,
    "dueAt": "2023-01-19T12:33:55.859Z",
    "memberPlan": "Foo Subscription",
    "paidAt": "2023-01-19T12:34:01.944Z",
  },
  {
    "amount": 1000,
    "dueAt": "2023-01-19T12:41:05.640Z",
    "memberPlan": "Foo Subscription",
    "paidAt": "2023-01-19T12:48:13.188Z",
  },
]
`;

exports[`DashboardInvoiceService should get the expected revenue of a month 2`] = `
{
  "include": {
    "items": true,
    "subscription": {
      "select": {
        "memberPlan": {
          "select": {
            "name": true,
          },
        },
      },
    },
  },
  "orderBy": {
    "dueAt": "desc",
  },
  "where": {
    "OR": [
      {
        "dueAt": {
          "gte": 2023-01-01T00:00:00.000Z,
          "lt": 2023-02-01T00:00:00.000Z,
        },
      },
      {
        "paidAt": {
          "gte": 2023-01-01T00:00:00.000Z,
          "lt": 2023-02-01T00:00:00.000Z,
        },
      },
    ],
    "canceledAt": null,
    "manuallySetAsPaidByUserId": null,
  },
}
`;

exports[`DashboardInvoiceService should get the revenue of a month 1`] = `
[
  {
    "amount": 150,
    "dueAt": "2023-01-19T12:41:05.640Z",
    "memberPlan": "Foo Subscription",
    "paidAt": "2023-01-19T12:48:13.188Z",
  },
  {
    "amount": 200,
    "dueAt": "2023-01-19T12:33:55.859Z",
    "memberPlan": "Foo Subscription",
    "paidAt": "2023-01-19T12:34:01.944Z",
  },
]
`;

exports[`DashboardInvoiceService should get the revenue of a month 2`] = `
{
  "include": {
    "items": {
      "select": {
        "amount": true,
        "quantity": true,
      },
    },
    "subscription": {
      "select": {
        "memberPlan": {
          "select": {
            "name": true,
          },
        },
      },
    },
  },
  "orderBy": {
    "paidAt": "desc",
  },
  "where": {
    "manuallySetAsPaidByUserId": null,
    "paidAt": {
      "gte": 2023-01-01T00:00:00.000Z,
      "lt": 2023-02-01T00:00:00.000Z,
    },
  },
}
`;
